src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs

```diff
@@ -36,7 +40,23 @@ public override void Initialize(AnalysisContext context)
             var wellKnownTypes = WellKnownTypes.GetOrCreate(compilation);
             var routeUsageCache = RouteUsageCache.GetOrCreate(compilation);
 
-            context.RegisterOperationAction(context =>
+            // Pool and reuse the same list for each block.
+            var concurrentQueue = new ConcurrentQueue<List<MapOperation>>();
+            context.RegisterOperationBlockStartAction(context =>
+            {
+                if (!concurrentQueue.TryDequeue(out var mapOperations))
+                {
+                    mapOperations = new List<MapOperation>();
+                }
+
+                context.RegisterOperationAction(c => DoOperationAnalysis(c, mapOperations), OperationKind.Invocation);
+                context.RegisterOperationBlockEndAction(c => DetectAmbiguousRoutes(c, mapOperations));
+
+                mapOperations.Clear();
+                concurrentQueue.Enqueue(mapOperations);
```

Comment:
This code will be executed in the OperationBlockStart callback itself - I presume you wanted it to be done at the end of OperationBlockEnd callback. I believe you want:

```csharp
context.RegisterOperationBlockEndAction(c =>
{
    DetectAmbiguousRoutes(c, mapOperations);

    mapOperations.Clear();
    concurrentQueue.Enqueue(mapOperations);
});
```